xapian-core/tests/unittest.cc

```diff
@@ -732,33 +741,34 @@ static bool test_movesupport1()
 	// Test move constructor
 	Xapian::Internal::intrusive_ptr_nonnull<A> p2(std::move(p1));
 	TEST_EQUAL(p2->get_x(), 5);
-	TEST_EQUAL(p1.get(), 0);
 
 	// Test move assignment
 	p3 = std::move(p2);
 	TEST_EQUAL(p3->get_x(), 5);
-	TEST_EQUAL(p2.get(), 0);
```

Comment:
Why did you remove these?

Checking `get()` is OK - it's just `_refs` that's not an implementation detail (and you still check `get()` elsewhere.

Reply:
Sorry, I forgot to mention this change. Thanks for pointing this.

Now test equality doesn't hold. 

It's clear in move constructor case - `TEST_EQUAL(p1.get(), 0)`, because `NULL` assignment not done.

####first version
I should've asked this before. I couldn't understand how equality worked first time in move assignment case - `TEST_EQUAL(p2.get(), 0)`. If move assignment really swapped, then `p2.get()` contain what p3 had before assignment. That didn't happen. How `p2.get()` gives 0?

I tried tracing in gdb. It calls assignment operator. It seems there is some optimization. Step through gdb was very confusing, I couldn't get much. Any advice how to debug?

####latest version
With the latest commit, for move assignment `TEST_EQUAL(p2.get(), 0)` doesn't hold, instead  `TEST_EQUAL(p2.get(), p3.get())` holds. It seems swap is not happening really.  Just one way assignment done.

Reply:
Oh I see - you had to change this because the move operations for the nonnull case now actually copy rather than move.  That's OK, but I think deserves a comment to explain why this differs in this way from the `intrusive_ptr` test above.

Not sure about the other question - I'll take a look.

Reply:
I think that's good - it's efficient and leaves `rhs` in a valid state, so it seems a valid approach.  And with the normal uses of `std:;move()` we should destroy `rhs` soon afterwards.